File: drivers/voltnuerongrid-driver-python/src/voltnuerongrid_driver_python/native_session.py
"""Minimal native wire session: Hello → Auth (optional) → Command (parity with TS `nativeSession.ts`)."""

from __future__ import annotations

import socket
from typing import Any, Dict, List, Optional

from .native_wire import encode_framed_json, read_framed_json


def native_command_roundtrip(
    host: str,
    port: int,
    session_id: str,
    command: str,
    body: Dict[str, Any],
    *,
    admin_api_key: Optional[str] = None,
    request_id_prefix: str = "py-native",
    connect_timeout_s: float = 5.0,
    idle_s: float = 30.0,
    max_frame_bytes: int = 1_048_576,
) -> Any:
    """General helper: TCP connect → Hello → Auth (optional) → Command → read Result → close socket."""
    addr = (host, port)
    rid = request_id_prefix
    with socket.create_connection(addr, timeout=connect_timeout_s) as sock:
        hello: Dict[str, Any] = {
            "frame_type": "Hello",
            "protocol_version": "v1",
            "request_id": f"{rid}-hello",
            "session_id": None,
            "payload": {
                "session_id": session_id,
                "protocol": "vng-native",
                "version": "v1",
            },
        }
        sock.sendall(encode_framed_json(hello))
        read_framed_json(sock, max_frame_bytes, idle_s)

        if admin_api_key:
            auth = {
                "frame_type": "Auth",
                "protocol_version": "v1",
                "request_id": f"{rid}-auth",
                "session_id": session_id,
                "payload": {"admin_api_key": admin_api_key},
            }
            sock.sendall(encode_framed_json(auth))
            read_framed_json(sock, max_frame_bytes, idle_s)

        payload: Dict[str, Any] = {"command": command}
        if body:
            payload["body"] = body

        cmd = {
            "frame_type": "Command",
            "protocol_version": "v1",
            "request_id": f"{rid}-cmd",
            "session_id": session_id,
            "payload": payload,
        }
        sock.sendall(encode_framed_json(cmd))
        return read_framed_json(sock, max_frame_bytes, idle_s)


def native_health_command_roundtrip(
    host: str,
    port: int,
    session_id: str,
    *,
    admin_api_key: Optional[str] = None,
    request_id_prefix: str = "py-native-health",
    connect_timeout_s: float = 5.0,
    idle_s: float = 30.0,
    max_frame_bytes: int = 1_048_576,
) -> Any:
    """Health command roundtrip — wraps native_command_roundtrip with command='health'."""
    return native_command_roundtrip(
        host,
        port,
        session_id,
        "health",
        {},
        admin_api_key=admin_api_key,
        request_id_prefix=request_id_prefix,
        connect_timeout_s=connect_timeout_s,
        idle_s=idle_s,
        max_frame_bytes=max_frame_bytes,
    )


def native_sql_execute_command_roundtrip(
    host: str,
    port: int,
    session_id: str,
    sql_batch: str,
    *,
    max_rows: Optional[int] = None,
    admin_api_key: Optional[str] = None,
    request_id_prefix: str = "py-native-sql-execute",
    connect_timeout_s: float = 5.0,
    idle_s: float = 30.0,
    max_frame_bytes: int = 1_048_576,
) -> Any:
    """SQL execute command roundtrip — command='sql.execute'."""
    body: Dict[str, Any] = {"sql_batch": sql_batch}
    if max_rows is not None:
        body["max_rows"] = max_rows
    return native_command_roundtrip(
        host,
        port,
        session_id,
        "sql.execute",
        body,
        admin_api_key=admin_api_key,
        request_id_prefix=request_id_prefix,
        connect_timeout_s=connect_timeout_s,
        idle_s=idle_s,
        max_frame_bytes=max_frame_bytes,
    )


def native_sql_analyze_command_roundtrip(
    host: str,
    port: int,
    session_id: str,
    sql_batch: str,
    *,
    admin_api_key: Optional[str] = None,
    request_id_prefix: str = "py-native-sql-analyze",
    connect_timeout_s: float = 5.0,
    idle_s: float = 30.0,
    max_frame_bytes: int = 1_048_576,
) -> Any:
    """SQL analyze command roundtrip — command='sql.analyze'."""
    return native_command_roundtrip(
        host,
        port,
        session_id,
        "sql.analyze",
        {"sql_batch": sql_batch},
        admin_api_key=admin_api_key,
        request_id_prefix=request_id_prefix,
        connect_timeout_s=connect_timeout_s,
        idle_s=idle_s,
        max_frame_bytes=max_frame_bytes,
    )


def native_sql_route_command_roundtrip(
    host: str,
    port: int,
    session_id: str,
    sql_batch: str,
    *,
    admin_api_key: Optional[str] = None,
    request_id_prefix: str = "py-native-sql-route",
    connect_timeout_s: float = 5.0,
    idle_s: float = 30.0,
    max_frame_bytes: int = 1_048_576,
) -> Any:
    """SQL route command roundtrip — command='sql.route'."""
    return native_command_roundtrip(
        host,
        port,
        session_id,
        "sql.route",
        {"sql_batch": sql_batch},
        admin_api_key=admin_api_key,
        request_id_prefix=request_id_prefix,
        connect_timeout_s=connect_timeout_s,
        idle_s=idle_s,
        max_frame_bytes=max_frame_bytes,
    )


def native_sql_transaction_command_roundtrip(
    host: str,
    port: int,
    session_id: str,
    statements: List[str],
    *,
    isolation_level: Optional[str] = None,
    admin_api_key: Optional[str] = None,
    request_id_prefix: str = "py-native-sql-transaction",
    connect_timeout_s: float = 5.0,
    idle_s: float = 30.0,
    max_frame_bytes: int = 1_048_576,
) -> Any:
    """SQL transaction command roundtrip — command='sql.transaction'."""
    body: Dict[str, Any] = {"statements": statements}
    if isolation_level is not None:
        body["isolation_level"] = isolation_level
    return native_command_roundtrip(
        host,
        port,
        session_id,
        "sql.transaction",
        body,
        admin_api_key=admin_api_key,
        request_id_prefix=request_id_prefix,
        connect_timeout_s=connect_timeout_s,
        idle_s=idle_s,
        max_frame_bytes=max_frame_bytes,
    )


def native_schema_registry_command_roundtrip(
    host: str,
    port: int,
    session_id: str,
    *,
    admin_api_key: Optional[str] = None,
    request_id_prefix: str = "py-native-schema-registry",
    connect_timeout_s: float = 5.0,
    idle_s: float = 30.0,
    max_frame_bytes: int = 1_048_576,
) -> Any:
    """Schema registry command roundtrip — command='ingest.schema.registry'."""
    return native_command_roundtrip(
        host,
        port,
        session_id,
        "ingest.schema.registry",
        {},
        admin_api_key=admin_api_key,
        request_id_prefix=request_id_prefix,
        connect_timeout_s=connect_timeout_s,
        idle_s=idle_s,
        max_frame_bytes=max_frame_bytes,
    )

File: drivers/voltnuerongrid-driver-typescript/src/nativeSession.ts
import * as net from "node:net";
import { encodeFramedJson, readFramedJson, FramedReader } from "./nativeWire";

export interface NativeHealthSessionOptions {
  host: string;
  port: number;
  sessionId: string;
  requestIdPrefix?: string;
  adminApiKey?: string;
  connectTimeoutMs?: number;
  idleMs?: number;
  maxFrameBytes?: number;
}

/** Extends base session options — no extra fields needed; kept for forward compatibility. */
export type NativeCommandSessionOptions = NativeHealthSessionOptions;

/**
 * Core session helper: TCP connect → Hello → Auth (optional) → send Command frame →
 * read Result frame → end socket.
 *
 * Returns the decoded Result payload JSON.
 */
export async function nativeCommandRoundtrip(
  opts: NativeCommandSessionOptions,
  command: string,
  body: Record<string, unknown>
): Promise<unknown> {
  const connectTimeoutMs = opts.connectTimeoutMs ?? 5000;
  const idleMs = opts.idleMs ?? 30_000;
  const maxFrameBytes = opts.maxFrameBytes ?? 1_048_576;
  const rid = opts.requestIdPrefix ?? `ts-native-${command}`;

  return new Promise((resolve, reject) => {
    let socket: net.Socket;
    const timer = setTimeout(() => {
      try {
        socket?.destroy();
      } catch {
        /* ignore */
      }
      reject(new Error(`native session connect timeout after ${connectTimeoutMs}ms`));
    }, connectTimeoutMs);

    socket = net.createConnection({ host: opts.host, port: opts.port }, async () => {
      clearTimeout(timer);
      // One FramedReader per connection — shares the buffer across all reads
      // so excess bytes delivered in a single TCP segment are never discarded.
      const reader = new FramedReader(socket);
      try {
        const hello = {
          frame_type: "Hello",
          protocol_version: "v1",
          request_id: `${rid}-hello`,
          session_id: null,
          payload: { session_id: opts.sessionId, protocol: "vng-native", version: "v1" },
        };
        socket.write(encodeFramedJson(hello));
        await readFramedJson(reader, maxFrameBytes, idleMs);

        if (opts.adminApiKey) {
          const auth = {
            frame_type: "Auth",
            protocol_version: "v1",
            request_id: `${rid}-auth`,
            session_id: opts.sessionId,
            payload: { admin_api_key: opts.adminApiKey },
          };
          socket.write(encodeFramedJson(auth));
          await readFramedJson(reader, maxFrameBytes, idleMs);
        }

        const cmd = {
          frame_type: "Command",
          protocol_version: "v1",
          request_id: `${rid}-cmd`,
          session_id: opts.sessionId,
          payload: { command, body },
        };
        socket.write(encodeFramedJson(cmd));
        const result = await readFramedJson(reader, maxFrameBytes, idleMs);
        socket.end();
        resolve(result);
      } catch (e) {
        socket.destroy();
        reject(e instanceof Error ? e : new Error(String(e)));
      }
    });

    socket.once("error", (e) => {
      clearTimeout(timer);
      reject(e);
    });
  });
}

/**
 * Minimal native wire session: TCP → Hello → Auth (optional) → Command `health` (Rust driver parity for probes).
 * Returns the decoded Result payload JSON.
 */
export async function nativeHealthCommandRoundtrip(
  opts: NativeHealthSessionOptions
): Promise<unknown> {
  const connectTimeoutMs = opts.connectTimeoutMs ?? 5000;
  const idleMs = opts.idleMs ?? 30_000;
  const maxFrameBytes = opts.maxFrameBytes ?? 1_048_576;
  const rid = opts.requestIdPrefix ?? "ts-native-health";

  return new Promise((resolve, reject) => {
    let socket: net.Socket;
    const timer = setTimeout(() => {
      try {
        socket?.destroy();
      } catch {
        /* ignore */
      }
      reject(new Error(`native session connect timeout after ${connectTimeoutMs}ms`));
    }, connectTimeoutMs);

    socket = net.createConnection({ host: opts.host, port: opts.port }, async () => {
      clearTimeout(timer);
      // One FramedReader per connection — shares the buffer across all reads.
      const reader = new FramedReader(socket);
      try {
        const hello = {
          frame_type: "Hello",
          protocol_version: "v1",
          request_id: `${rid}-hello`,
          session_id: null,
          payload: { session_id: opts.sessionId, protocol: "vng-native", version: "v1" },
        };
        socket.write(encodeFramedJson(hello));
        await readFramedJson(reader, maxFrameBytes, idleMs);

        if (opts.adminApiKey) {
          const auth = {
            frame_type: "Auth",
            protocol_version: "v1",
            request_id: `${rid}-auth`,
            session_id: opts.sessionId,
            payload: { admin_api_key: opts.adminApiKey },
          };
          socket.write(encodeFramedJson(auth));
          await readFramedJson(reader, maxFrameBytes, idleMs);
        }

        const cmd = {
          frame_type: "Command",
          protocol_version: "v1",
          request_id: `${rid}-cmd`,
          session_id: opts.sessionId,
          payload: { command: "health" },
        };
        socket.write(encodeFramedJson(cmd));
        const result = await readFramedJson(reader, maxFrameBytes, idleMs);
        socket.end();
        resolve(result);
      } catch (e) {
        socket.destroy();
        reject(e instanceof Error ? e : new Error(String(e)));
      }
    });

    socket.once("error", (e) => {
      clearTimeout(timer);
      reject(e);
    });
  });
}

/** Body type for sql.execute command. */
export interface SqlExecuteBody {
  sql_batch: string;
  max_rows?: number;
}

/**
 * Native session roundtrip for `sql.execute`.
 * Sends Command frame with `command: "sql.execute"` and `body: { sql_batch, max_rows? }`.
 */
export function nativeSqlExecuteCommandRoundtrip(
  opts: NativeCommandSessionOptions,
  body: SqlExecuteBody
): Promise<unknown> {
  const frameBody: Record<string, unknown> = { sql_batch: body.sql_batch };
  if (body.max_rows !== undefined) {
    frameBody.max_rows = body.max_rows;
  }
  return nativeCommandRoundtrip(opts, "sql.execute", frameBody);
}

/** Body type for sql.analyze command. */
export interface SqlAnalyzeBody {
  sql_batch: string;
}

/**
 * Native session roundtrip for `sql.analyze`.
 * Sends Command frame with `command: "sql.analyze"` and `body: { sql_batch }`.
 */
export function nativeSqlAnalyzeCommandRoundtrip(
  opts: NativeCommandSessionOptions,
  body: SqlAnalyzeBody
): Promise<unknown> {
  return nativeCommandRoundtrip(opts, "sql.analyze", { sql_batch: body.sql_batch });
}

/** Body type for sql.route command. */
export interface SqlRouteBody {
  sql_batch: string;
}

/**
 * Native session roundtrip for `sql.route`.
 * Sends Command frame with `command: "sql.route"` and `body: { sql_batch }`.
 */
export function nativeSqlRouteCommandRoundtrip(
  opts: NativeCommandSessionOptions,
  body: SqlRouteBody
): Promise<unknown> {
  return nativeCommandRoundtrip(opts, "sql.route", { sql_batch: body.sql_batch });
}

/** Body type for sql.transaction command. */
export interface SqlTransactionBody {
  statements: string[];
  isolation_level?: string;
}

/**
 * Native session roundtrip for `sql.transaction`.
 * Sends Command frame with `command: "sql.transaction"` and `body: { statements, isolation_level? }`.
 */
export function nativeSqlTransactionCommandRoundtrip(
  opts: NativeCommandSessionOptions,
  body: SqlTransactionBody
): Promise<unknown> {
  const frameBody: Record<string, unknown> = { statements: body.statements };
  if (body.isolation_level !== undefined) {
    frameBody.isolation_level = body.isolation_level;
  }
  return nativeCommandRoundtrip(opts, "sql.transaction", frameBody);
}

/**
 * Native session roundtrip for `ingest.schema.registry`.
 * Sends Command frame with `command: "ingest.schema.registry"` and empty body `{}`.
 */
export function nativeSchemaRegistryCommandRoundtrip(
  opts: NativeCommandSessionOptions
): Promise<unknown> {
  return nativeCommandRoundtrip(opts, "ingest.schema.registry", {});
}
